<!DOCTYPE html>
<html lang="en   ">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <title>OOMPA CloneFinder Packages</title>
    <link href="https://oompa.r-forge.r-project.org/bianchi.css" rel="stylesheet" type="text/css" />
  </head>
<body>

<!-- <img src="tuba3.gif" alt="OOMPA logo" width="262px"  height="156px" /> -->
<div id="logobar">
  <img src="https://oompa.r-forge.r-project.org/tuba2.gif" alt="OOMPA logo" height="262px" width="156px" />
<br />
<br />
<a href="https://oompa.r-forge.r-project.org/index.html">OOMPA Overview</a>
<br />
</div>
<div id="mainarea">
<h1>OOMPA PROJECT</h1>
<h2>CloneFinder</h2>
The CloneFinder
<A HREF="https://r-forge.r-project.org/">R-Forge</A> project is one
component of the broader OOMPA Project produced by the
<A HREF = "https://gitlab.com/krcoombes/coombeslab">Coombes Lab</A>
over the past decade. it consists of a set of packages to understand
genetic structural variations and how the evolution of tumors gives
rise to distinct clones.

<h3>Support</h3>
Support for all packages that are part of the OOMPA family uses a
common set of discussion forums and bug trackers:
<ul>
<li><a href="https://r-forge.r-project.org/forum/forum.php?forum_id=4515&group_id=1746">
    User Help Forum</a></li>
<li><a href="https://r-forge.r-project.org/forum/forum.php?forum_id=4516&group_id=1746">
    Developer Forum</a></li>
<li><a href="https://r-forge.r-project.org/tracker/?atid=5395&group_id=1746&func=browse">
    Feature Requests</a></li>
<li><a href="https://r-forge.r-project.org/tracker/?atid=5394&group_id=1746&func=browse">
    Bug Reports</a> </li>
</ul>


<h3>List of Packages</h3>
<dl>
<dt><A HREF="http://clonefinder.r-forge.r-project.org/packages/CloneSeeker.html">CloneSeeker<A></dt>
<dd>The CloneSeeker package was developed as part of Mark Zucker's PhD
  thesis, and its description can be found in the paper
  by <A HREF="https://pubmed.ncbi.nlm.nih.gov/31161200/">Zucker and
  colleagues</A>. The method uses bulk cytogenetic data (arising, for
  example, from the analysis of SNP-chip or comparative genomic
  hybridization data) to reconstruct the clonal architecture of a
  specific tumor. It can also use mutation allele frequency data
  (with or without the cytogentic structural variaiton data) for the
  same purpose.</dd>
<dt><A HREF="http://clonefinder.r-forge.r-project.org/packages/CloneData.html">CloneData<A></dt>
  <dd>CloneData is a data package, containing the data that was
  simulated to support Mark Zucker's PhD thesis and his paper in
  Bioinformatics. It also includes the code to generate those simulations.</dd>
<dt><A HREF="http://clonefinder.r-forge.r-project.org/packages/SVAlignR.html">SVAlignR<A></dt>
  <dd>The SVAlignR package is still under development, which is why
  it has not yet been submitted to CRAN. The goal is to study
  long-read sequences (such as those from Oxford Nanopore or Pacific
  Biosystems) from virus-associated tumors in order to determine if
  the virus is actually incorporated into the host genome or if it
  merely exists as extrachromosomal units that possibly include
  copies of pieces of the host genome.</dd>
</dl>




<BR />
<BR />
<!-- R-Forge Logo -->
<table border="0" width="100%" cellspacing="0" cellpadding="0">
<tr><td>
<a href="http://r-forge.r-project.org/">
 <img src="http://r-forge.r-project.org/themes/rforge//imagesrf/logo.png" 
      border="0" alt="R-Forge Logo" /> </a> </td> </tr>
</table>

</div> <!-- end mainarea div-->


</body>
</html>
